ethereum/rpc/namespaces/miner/api.go

```diff
@@ -0,0 +1,163 @@
+}
+
+// NewMinerAPI creates an instance of the Miner API.
+func NewMinerAPI(
+	ctx *server.Context,
+	ethAPI *eth.PublicAPI,
+) *API {
+	return &API{
+		ctx:    ctx,
+		ethAPI: ethAPI,
+		logger: ctx.Logger.With("module", "miner"),
+	}
+}
+
+// SetEtherbase sets the etherbase of the miner
+func (api *API) SetEtherbase(etherbase common.Address) bool {
+	api.logger.Debug("miner_setEtherbase")
+
+	list, err := api.ethAPI.ClientCtx().Keyring.List()
+	if err != nil || len(list) < 1 {
+		api.logger.Debug("Could not get list of addresses")
+		return false
+	}
+
+	addr := common.BytesToAddress(list[0].GetPubKey().Address())
+
+	api.logger.Info(addr.String())
+
+	args := rpctypes.SendTxArgs{}
+	args.From = addr
+	// TODO: set this as the message info
+	// delegatorAddress := addr
+	// withdrawAddress := etherbase
+	args.Data = &hexutil.Bytes{}
+	args, err = api.ethAPI.SetTxDefaults(args)
+	if err != nil {
+		api.logger.Debug("Unable to parse transaction args", "error", err.Error())
+		return false
+	}
+
+	msg := args.ToTransaction()
+
+	if err := msg.ValidateBasic(); err != nil {
+		api.logger.Debug("tx failed basic validation", "error", err.Error())
+		return false
+	}
+
+	signer := ethtypes.LatestSignerForChainID(args.ChainID.ToInt())
+
+	// Sign transaction
+	if err := msg.Sign(signer, api.ethAPI.ClientCtx().Keyring); err != nil {
+		api.logger.Debug("failed to sign tx", "error", err.Error())
+		return false
+	}
+
+	// Assemble transaction from fields
+	builder, ok := api.ethAPI.ClientCtx().TxConfig.NewTxBuilder().(authtx.ExtensionOptionsTxBuilder)
+	if !ok {
+		api.logger.Error("clientCtx.TxConfig.NewTxBuilder returns unsupported builder", "error", err.Error())
+	}
+
+	option, err := codectypes.NewAnyWithValue(&evmtypes.ExtensionOptionsEthereumTx{})
+	if err != nil {
+		api.logger.Error("codectypes.NewAnyWithValue failed to pack an obvious value", "error", err.Error())
+		return false
+	}
+
+	builder.SetExtensionOptions(option)
+	err = builder.SetMsgs(msg)
+	if err != nil {
+		api.logger.Error("builder.SetMsgs failed", "error", err.Error())
+	}
+
+	// Query params to use the EVM denomination
+	res, err := api.ethAPI.QueryClient().QueryClient.Params(api.ethAPI.Ctx(), &evmtypes.QueryParamsRequest{})
+	if err != nil {
+		api.logger.Error("failed to query evm params", "error", err.Error())
+		return false
+	}
+
+	txData, err := evmtypes.UnpackTxData(msg.Data)
+	if err != nil {
+		api.logger.Error("failed to unpack tx data", "error", err.Error())
+		return false
+	}
+
+	fees := sdk.Coins{sdk.NewCoin(res.Params.EvmDenom, sdk.NewIntFromBigInt(txData.Fee()))}
+	builder.SetFeeAmount(fees)
+	builder.SetGasLimit(msg.GetGas())
+
+	// Encode transaction by default Tx encoder
+	txEncoder := api.ethAPI.ClientCtx().TxConfig.TxEncoder()
+	txBytes, err := txEncoder(builder.GetTx())
+	if err != nil {
+		api.logger.Error("failed to encode eth tx using default encoder", "error", err.Error())
+		return false
+	}
+
+	txHash := msg.AsTransaction().Hash()
```

Comment:
the hash is calculated differently for non evm txs. use instead the tendermint library:

```
tmHash :=  `common.BytesToHash(tmtypes.Tx(txBytes).Hash())`
```